File: web/routes/pro_analysis.py
"""
🎯 PRO ANALYSIS API - Blueprint

Endpoint pour l'analyse technique professionnelle détaillée.
Utilisé par le panneau "Analyse Pro" de l'interface web.
"""

import logging
from flask import Blueprint, jsonify

logger = logging.getLogger(__name__)

pro_analysis_bp = Blueprint('pro_analysis', __name__)

# Import conditionnel des analyseurs
try:
    from dashboard.technical_analysis import TechnicalAnalyzer
    TECHNICAL_ANALYZER_AVAILABLE = True
except ImportError:
    TECHNICAL_ANALYZER_AVAILABLE = False
    logger.warning("⚠️  TechnicalAnalyzer non disponible pour pro_analysis")

# 🔥 TEMPORAIREMENT DÉSACTIVÉ : ProTechnicalAnalyzer attend un DataFrame, pas un symbol
# TODO: Adapter ProTechnicalAnalyzer.analyze(symbol, period) au lieu de analyze(df)
PRO_ANALYZER_AVAILABLE = False
pro_analyzer_instance = None

# try:
#     from web.utils.pro_technical_analyzer import ProTechnicalAnalyzer
#     PRO_ANALYZER_AVAILABLE = True
#     pro_analyzer_instance = ProTechnicalAnalyzer()
# except ImportError:
#     PRO_ANALYZER_AVAILABLE = False
#     pro_analyzer_instance = None
# except Exception as e:
#     PRO_ANALYZER_AVAILABLE = False
#     pro_analyzer_instance = None
#     logger.warning(f"⚠️  ProTechnicalAnalyzer init error: {e}")


@pro_analysis_bp.route('/api/pro-analysis/<symbol>')
def api_pro_analysis(symbol):
    """
    🎯 Analyse technique professionnelle détaillée
    
    Args:
        symbol: Ticker du symbole (ex: INTC, AAPL)
    
    Returns:
        JSON avec analyse complète :
        - overall_signal: STRONG_BUY, BUY, HOLD, SELL, STRONG_SELL
        - confidence: Niveau de confiance (0-100)
        - risk_level: LOW, MEDIUM, HIGH
        - current_price: Prix actuel
        - trend: Analyse de tendance
        - momentum: RSI et momentum
        - macd: MACD analysis
        - volatility: Bollinger Bands
        - volume: OBV analysis
        - trading_plan: Plan de trading textuel
    """
    
    # Version simplifiée avec TechnicalAnalyzer
    if not TECHNICAL_ANALYZER_AVAILABLE:
        return jsonify({
            'error': 'Pro Analysis indisponible',
            'details': 'Aucun analyseur technique disponible'
        }), 503
    
    try:
        # Analyse avec TechnicalAnalyzer (période 3 mois)
        analyzer = TechnicalAnalyzer(symbol, period='3mo', interval='1d')
        df = analyzer.df
        indicators = analyzer.get_all_indicators()
        signal = analyzer.generate_signal()
        
        # Prix actuel
        current_price = float(df['Close'].iloc[-1])
        
        # Valeurs par défaut pour stop_loss/take_profit si None
        entry_price = signal.entry_price if signal.entry_price is not None else current_price
        stop_loss = signal.stop_loss if signal.stop_loss is not None else current_price * 0.98  # -2%
        take_profit = signal.take_profit if signal.take_profit is not None else current_price * 1.03  # +3%
        
        # === RSI ANALYSIS ===
        rsi_value = float(indicators.get('rsi', 50.0))
        if rsi_value > 70:
            rsi_zone = 'overbought'
            rsi_signal = 'SELL'
        elif rsi_value < 30:
            rsi_zone = 'oversold'
            rsi_signal = 'BUY'
        else:
            rsi_zone = 'neutral'
            rsi_signal = 'HOLD'
        
        # === MACD ANALYSIS ===
        macd_data = indicators.get('macd', {})
        if isinstance(macd_data, dict):
            macd_value = float(macd_data.get('macd', 0.0))
            macd_signal_value = float(macd_data.get('signal', 0.0))
            macd_histogram = float(macd_data.get('histogram', 0.0))
        else:
            macd_value = float(macd_data) if macd_data else 0.0
            macd_signal_value = 0.0
            macd_histogram = 0.0
        
        macd_signal = 'BUY' if macd_value > macd_signal_value else 'SELL'
        macd_crossover = None
        if macd_value > macd_signal_value and macd_histogram > 0:
            macd_crossover = 'bullish'
        elif macd_value < macd_signal_value and macd_histogram < 0:
            macd_crossover = 'bearish'
        
        # === BOLLINGER BANDS ===
        bb_data = indicators.get('bollinger_bands', {})
        if isinstance(bb_data, dict):
            bb_upper = float(bb_data.get('upper', current_price * 1.02))
            bb_middle = float(bb_data.get('middle', current_price))
            bb_lower = float(bb_data.get('lower', current_price * 0.98))
        else:
            bb_upper = float(indicators.get('bb_upper', current_price * 1.02))
            bb_middle = float(indicators.get('bb_middle', current_price))
            bb_lower = float(indicators.get('bb_lower', current_price * 0.98))
        
        if bb_middle > 0:
            bb_width = ((bb_upper - bb_lower) / bb_middle * 100)
        else:
            bb_width = 0
        
        if current_price > bb_middle:
            price_position = 'upper'
        elif current_price < bb_middle:
            price_position = 'lower'
        else:
            price_position = 'middle'
        
        squeeze_detected = bb_width < 2.0
        
        # === SIGNAL GLOBAL ===
        if signal.signal == 'BUY':
            overall_signal = 'STRONG_BUY' if signal.strength > 70 else 'BUY'
        elif signal.signal == 'SELL':
            overall_signal = 'STRONG_SELL' if signal.strength > 70 else 'SELL'
        else:
            overall_signal = 'HOLD'
        
        # === RISQUE ===
        if bb_width > 5 or signal.strength < 50:
            risk_level = 'HIGH'
        elif bb_width > 2.5 or signal.strength < 70:
            risk_level = 'MEDIUM'
        else:
            risk_level = 'LOW'
        
        # === SMA 200 ===
        sma_200 = float(indicators.get('sma_200', current_price))
        price_vs_sma200 = 'above' if current_price > sma_200 else 'below'
        
        sma_50 = float(indicators.get('sma_50', current_price))
        golden_cross = sma_50 > sma_200
        
        # === ADX (Force de tendance) ===
        adx_value = float(indicators.get('adx', 0.0))
        
        # === PLAN DE TRADING ===
        stop_pct = ((stop_loss - current_price) / current_price * 100)
        target_pct = ((take_profit - current_price) / current_price * 100)
        
        trading_plan = f"""SIGNAL: {overall_signal}

🎯 ENTRÉE: ${entry_price:.2f}
🛑 STOP LOSS: ${stop_loss:.2f} ({stop_pct:.1f}%)
🎯 TAKE PROFIT: ${take_profit:.2f} ({target_pct:.1f}%)

⚠️ RISQUE: {risk_level}
🎯 CONFIANCE: {signal.confidence:.0f}%
📊 ADX: {adx_value:.1f}

📊 RAISONS:
""" + '\n'.join(f"  • {r}" for r in signal.reasons)
        
        # === RÉPONSE JSON ===
        return jsonify({
            'overall_signal': overall_signal,
            'confidence': signal.confidence,
            'risk_level': risk_level,
            'current_price': current_price,
            
            'trend': {
                'direction': signal.trend,
                'strength': signal.strength,
                'explanation': f"Tendance {signal.trend.lower()} détectée avec force {signal.strength:.0f}%",
                'price_vs_sma200': price_vs_sma200,
                'golden_cross': golden_cross,
                'support_level': stop_loss,
                'resistance_level': take_profit,
                'adx': adx_value
            },
            
            'momentum': {
                'signal': rsi_signal,
                'rsi_value': rsi_value,
                'zone': rsi_zone,
                'explanation': f"RSI à {rsi_value:.1f} - Zone {rsi_zone}",
                'divergence_detected': False,
                'divergence_type': None
            },
            
            'macd': {
                'signal': macd_signal,
                'macd_value': macd_value,
                'signal_value': macd_signal_value,
                'histogram_value': macd_histogram,
                'crossover': macd_crossover,
                'histogram_direction': 'increasing' if macd_histogram > 0 else 'decreasing',
                'explanation': f"MACD {'au-dessus' if macd_value > macd_signal_value else 'en-dessous'} du signal"
            },
            
            'volatility': {
                'price_position': price_position,
                'bb_width': bb_width,
                'bb_upper': bb_upper,
                'bb_middle': bb_middle,
                'bb_lower': bb_lower,
                'squeeze_detected': squeeze_detected,
                'explanation': f"Volatilité {risk_level.lower()} - Largeur BB: {bb_width:.2f}%"
            },
            
            'volume': {
                'obv_trend': signal.trend,
                'volume_confirmation': True,
                'smart_money_accumulation': signal.signal == 'BUY',
                'explanation': f"Volume confirme la tendance {signal.trend.lower()}"
            },
            
            'trading_plan': trading_plan
        })
        
    except Exception as e:
        logger.error(f"❌ Erreur pro-analysis {symbol}: {e}")
        import traceback
        logger.error(traceback.format_exc())
        return jsonify({
            'error': str(e),
            'details': 'Erreur lors de l\'analyse du symbole'
        }), 500

File: web/static/pro_analysis.js
/**
 * 🎯 PRO ANALYSIS - Module Frontend
 * 
 * Affiche l'analyse technique professionnelle dans le dashboard
 * Utilise l'API /api/pro-analysis/<ticker>
 */

class ProAnalysis {
    constructor() {
        this.currentData = null;
        this.currentTicker = null;
        this.isLoading = false;
        this.cache = {}; // Cache court (2 min)
        this.refreshInterval = null;
        this.initEventListeners();
    }
    
    /**
     * 🔄 Charger l'analyse pour un ticker
     */
    async loadAnalysis(ticker, forceRefresh = false) {
        // ⚠️ ANTI-DOUBLON : Si déjà en cours, skip
        if (this.isLoading) {
            console.log(`⏭️ Skip pro-analysis ${ticker} (chargement en cours)`);
            return;
        }
        
        // 📂 Vérifier le cache (valide 2 min seulement)
        if (!forceRefresh && this.cache[ticker] && (Date.now() - this.cache[ticker].timestamp < 120000)) {
            console.log(`💾 Cache pro-analysis ${ticker}`);
            this.currentData = this.cache[ticker].data;
            this.currentTicker = ticker;
            this.renderSummary();
            this.renderDetails();
            return;
        }
        
        try {
            this.isLoading = true;
            console.log(`🎯 ${forceRefresh ? '🔄 Rafraîchissement' : 'Chargement'} analyse pro pour ${ticker}...`);
            
            const response = await fetch(`/api/pro-analysis/${ticker}`);
            
            if (!response.ok) {
                throw new Error(`Erreur HTTP ${response.status}`);
            }
            
            this.currentData = await response.json();
            this.currentTicker = ticker;
            
            // 💾 Mise en cache (2 min)
            this.cache[ticker] = {
                data: this.currentData,
                timestamp: Date.now()
            };
            
            console.log('✅ Analyse pro:', this.currentData.overall_signal, this.currentData.confidence + '%');
            
            this.renderSummary();
            this.renderDetails();
            
            // 🔄 Démarrer l'auto-refresh si pas déjà actif
            this.startAutoRefresh();
            
        } catch (error) {
            console.error('❌ Erreur chargement analyse pro:', error);
            this.renderError(error.message);
        } finally {
            this.isLoading = false;
        }
    }
    
    /**
     * 🔄 Auto-refresh toutes les 2 minutes
     */
    startAutoRefresh() {
        // Arrêter l'ancien interval si existe
        if (this.refreshInterval) {
            clearInterval(this.refreshInterval);
        }
        
        // Nouveau refresh toutes les 2 minutes (120 000 ms)
        this.refreshInterval = setInterval(() => {
            if (this.currentTicker && !this.isLoading) {
                console.log(`🔄 Auto-refresh pro-analysis ${this.currentTicker}`);
                this.loadAnalysis(this.currentTicker, true); // forceRefresh = true
            }
        }, 120000); // 2 minutes
        
        console.log('⏰ Auto-refresh activé (toutes les 2 min)');
    }
    
    /**
     * ⏸️ Arrêter l'auto-refresh
     */
    stopAutoRefresh() {
        if (this.refreshInterval) {
            clearInterval(this.refreshInterval);
            this.refreshInterval = null;
            console.log('⏸️ Auto-refresh désactivé');
        }
    }
    
    /**
     * 📊 Afficher le résumé (sidebar gauche)
     */
    renderSummary() {
        const container = document.getElementById('pro-summary');
        if (!container || !this.currentData) return;
        
        const data = this.currentData;
        const signalClass = `signal-${data.overall_signal.replace('_', '-')}`;
        const riskClass = `risk-${data.risk_level}`;
        
        // Afficher l'âge du cache
        const cacheAge = this.cache[this.currentTicker] ? 
            Math.floor((Date.now() - this.cache[this.currentTicker].timestamp) / 1000) : 0;
        
        container.innerHTML = `
            <div class="signal-badge ${signalClass} mb-3">
                ${this.getSignalEmoji(data.overall_signal)} ${data.overall_signal.replace('_', ' ')}
            </div>
            
            <div class="mb-3">
                <div class="text-gray-400 text-xs mb-1">Confiance</div>
                <div class="gauge">
                    <div class="gauge-fill" style="width: ${data.confidence}%; background: linear-gradient(90deg, #3b82f6, #8b5cf6);"></div>
                </div>
                <div class="text-right text-xs mt-1">${data.confidence.toFixed(0)}%</div>
            </div>
            
            <div class="mb-3">
                <div class="text-gray-400 text-xs">Risque</div>
                <div class="${riskClass} font-bold text-lg">${data.risk_level}</div>
            </div>
            
            <div class="text-xs bg-gray-800 p-2 rounded mt-3">
                <div class="text-gray-400 mb-1">Prix actuel</div>
                <div class="text-xl font-bold">${data.current_price.toFixed(2)} $</div>
                <div class="text-gray-500 text-xs mt-1">🔄 il y a ${cacheAge}s</div>
            </div>
        `;
    }
    
    /**
     * 📑 Afficher les détails (sidebar droite)
     */
    renderDetails() {
        const container = document.getElementById('pro-details');
        if (!container || !this.currentData) return;
        
        const data = this.currentData;
        
        let html = '';
        
        // 📈 TENDANCE
        html += this.renderIndicatorCard(
            'Tendance',
            'trend',
            data.trend.direction,
            data.trend.strength,
            data.trend.explanation,
            {
                'SMA 200': data.trend.price_vs_sma200 === 'above' ? '🟢 Au-dessus' : '🔴 En-dessous',
                'Golden Cross': data.trend.golden_cross ? '✅ Oui' : '❌ Non',
                'Support': data.trend.support_level ? `$${data.trend.support_level.toFixed(2)}` : 'N/A',
                'Résistance': data.trend.resistance_level ? `$${data.trend.resistance_level.toFixed(2)}` : 'N/A'
            }
        );
        
        // ⚡ MOMENTUM
        html += this.renderIndicatorCard(
            'Momentum (RSI)',
            'momentum',
            data.momentum.signal,
            data.momentum.rsi_value,
            data.momentum.explanation,
            {
                'RSI': `${data.momentum.rsi_value.toFixed(1)}`,
                'Zone': this.getRSIZoneEmoji(data.momentum.zone) + ' ' + data.momentum.zone,
                'Divergence': data.momentum.divergence_detected ? `⚠️ ${data.momentum.divergence_type}` : '✅ Non'
            }
        );
        
        // 📉 MACD
        html += this.renderIndicatorCard(
            'MACD',
            'macd',
            data.macd.signal,
            data.macd.macd_value,
            data.macd.explanation,
            {
                'MACD': data.macd.macd_value.toFixed(3),
                'Signal': data.macd.signal_value.toFixed(3),
                'Histogram': data.macd.histogram_value.toFixed(3),
                'Croisement': data.macd.crossover ? `🔀 ${data.macd.crossover}` : '➖ Aucun',
                'Direction': this.getDirectionEmoji(data.macd.histogram_direction)
            }
        );
        
        // 🌊 VOLATILITÉ
        html += this.renderIndicatorCard(
            'Volatilité (Bollinger)',
            'volatility',
            data.volatility.price_position,
            data.volatility.bb_width,
            data.volatility.explanation,
            {
                'BB Upper': `$${data.volatility.bb_upper.toFixed(2)}`,
                'BB Middle': `$${data.volatility.bb_middle.toFixed(2)}`,
                'BB Lower': `$${data.volatility.bb_lower.toFixed(2)}`,
                'Squeeze': data.volatility.squeeze_detected ? '💥 OUI !' : '❌ Non'
            }
        );
        
        // 📦 VOLUME
        html += this.renderIndicatorCard(
            'Volume (OBV)',
            'volume',
            data.volume.obv_trend,
            0,
            data.volume.explanation,
            {
                'Tendance OBV': this.getDirectionEmoji(data.volume.obv_trend),
                'Confirmation': data.volume.volume_confirmation ? '✅ Oui' : '⚠️ Non',
                'Smart Money': data.volume.smart_money_accumulation ? '👑 Détecté' : '❌ Non'
            }
        );
        
        // 📄 PLAN DE TRADING
        html += `
            <div class="bg-gray-800 p-3 rounded-lg mt-3">
                <div class="font-bold text-sm mb-2">📌 Plan de Trading</div>
                <div class="text-xs whitespace-pre-wrap">${data.trading_plan}</div>
            </div>
        `;
        
        container.innerHTML = html;
    }
    
    /**
     * 🎴 Créer une carte indicateur
     */
    renderIndicatorCard(title, type, signal, value, explanation, details) {
        let detailsHtml = '';
        for (const [key, val] of Object.entries(details)) {
            detailsHtml += `
                <div class="flex justify-between text-xs mb-1">
                    <span class="text-gray-400">${key}:</span>
                    <span class="font-semibold">${val}</span>
                </div>
            `;
        }
        
        return `
            <div class="indicator-card ${type}">
                <div class="font-bold text-sm mb-2">${title}</div>
                <div class="text-xs text-gray-300 mb-2">${explanation}</div>
                ${detailsHtml}
            </div>
        `;
    }
    
    /**
     * 🚫 Afficher une erreur
     */
    renderError(message) {
        const containers = ['pro-summary', 'pro-details'];
        containers.forEach(id => {
            const el = document.getElementById(id);
            if (el) {
                el.innerHTML = `
                    <div class="text-red-400 text-center py-4">
                        ❌ Erreur<br>
                        <span class="text-xs">${message}</span>
                    </div>
                `;
            }
        });
    }
    
    /**
     * 🧹 Nettoyer le cache (appel manuel si besoin)
     */
    clearCache() {
        this.cache = {};
        console.log('🧹 Cache Pro Analysis vidé');
    }
    
    /**
     * 📦 Helpers
     */
    getSignalEmoji(signal) {
        const emojis = {
            'STRONG_BUY': '🟢🚀',
            'BUY': '🟢',
            'HOLD': '🟡',
            'SELL': '🔴',
            'STRONG_SELL': '🔴⚠️'
        };
        return emojis[signal] || '🔵';
    }
    
    getRSIZoneEmoji(zone) {
        const emojis = {
            'oversold': '🔵',
            'neutral': '⚪',
            'overbought': '🔴'
        };
        return emojis[zone] || '⚪';
    }
    
    getDirectionEmoji(direction) {
        if (direction.includes('increas') || direction === 'rising') return '📈';
        if (direction.includes('decreas') || direction === 'falling') return '📉';
        return '➡️';
    }
    
    /**
     * 🎯 Event Listeners
     */
    initEventListeners() {
        // Auto-reload quand un nouveau ticker est analysé
        window.addEventListener('chartDataLoaded', (event) => {
            if (event.detail && event.detail.ticker) {
                // Délai de 500ms pour éviter les appels simultanés
                setTimeout(() => {
                    this.loadAnalysis(event.detail.ticker);
                }, 500);
            }
        });
        
        // Arrêter l'auto-refresh quand l'onglet est inactif (optimisation)
        document.addEventListener('visibilitychange', () => {
            if (document.hidden) {
                console.log('💤 Onglet inactif - pause auto-refresh');
                this.stopAutoRefresh();
            } else {
                console.log('👁️ Onglet actif - reprise auto-refresh');
                if (this.currentTicker) {
                    this.startAutoRefresh();
                }
            }
        });
    }
}

// 🚀 Auto-init
if (typeof window.ProAnalysis === 'undefined') {
    window.ProAnalysis = new ProAnalysis();
    console.log('✅ ProAnalysis module loaded');
}
